### PR TITLE
fix(consul): fix consul discovery only fetch healthy endpoint

### PR DESCRIPTION
**Pull request by dyrnq**
### Description

 Fixes https://github.com/apache/apisix/issues/8856

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/discovery/consul/init.lua [hunk 1]
```diff
@@ -215,14 +215,38 @@ function _M.connect(premature, consul_server, retry_delay)
...
+        local catalog_result, catalog_err = consul_client_catalog:get(consul_server.consul_catalog_sub_url)
+        local catalog_error_info = (catalog_err ~= nil and catalog_err)
+                or ((catalog_result ~= nil and catalog_result.status ~= 200)
+                and catalog_result.status)
+        if catalog_error_info then
+            log.error("connect consul: ", consul_server.consul_server_url,
+                " by sub url: ", consul_server.consul_catalog_sub_url,
+                ", got watch result: ", json_delay_encode(catalog_result, true),
+                ", with error: ", catalog_error_info)
+            -- retry_delay = get_retry_delay(retry_delay)
```
Comment (Fabriceli) [inline apisix/discovery/consul/init.lua:235]: The commented code should be removed. if it should be kept, please give the reason and put the reason in the comment

**Comment by soulbird** [comment]
Please add test cases

File: apisix/discovery/consul/init.lua [hunk 2]
```diff
@@ -246,7 +270,7 @@ function _M.connect(premature, consul_server, retry_delay)
                 -- add services to table
                 local nodes = up_services[service_name]
                 for  _, node in ipairs(result.body) do
-                    local svc_address, svc_port = node.ServiceAddress, node.ServicePort
+                    local svc_address, svc_port = node.Service.Address, node.Service.Port
```
Comment (Fabriceli) [inline apisix/discovery/consul/init.lua:273]: need to check whether `node.Service` is nil

**Review by Fabriceli — CHANGES_REQUESTED** [review]

File: apisix/discovery/consul/init.lua (hunk 1 above)
Comment (dyrnq) [reply apisix/discovery/consul/init.lua]: removed done

File: apisix/discovery/consul/init.lua (hunk 2 above)
Comment (dyrnq) [reply apisix/discovery/consul/init.lua]: checked done

**Comment by dyrnq** [comment]
> Please add test cases

Considering that this is a fix, there are some test case before this,  can we use the previous test case?

**Comment by dyrnq** [comment]
@soulbird test case added

File: t/discovery/consul.t [hunk 3]
```diff
@@ -576,3 +576,23 @@ GET /thc
 [{"host":"127.0.0.1","port":30513,"priority":0,"weight":1},{"host":"127.0.0.1","port":30514,"priority":0,"weight":1}]
 [{"host":"127.0.0.1","port":30513,"priority":0,"weight":1},{"host":"127.0.0.1","port":30514,"priority":0,"weight":1}]
 --- ignore_error_log
+
+
+
+=== TEST 13: prepare healthy and unhealthy nodes (with consul health check)
```
Comment (soulbird) [inline t/discovery/consul.t:582]: This is just to register or unregister some nodes with consul, so what will happen after using consul service discovery on apisix?

Comment (dyrnq) [reply t/discovery/consul.t:582]: I change use consul_dump.t test this

File: apisix/discovery/consul/init.lua [hunk 4]
```diff
@@ -246,7 +267,10 @@ function _M.connect(premature, consul_server, retry_delay)
                 -- add services to table
                 local nodes = up_services[service_name]
                 for  _, node in ipairs(result.body) do
-                    local svc_address, svc_port = node.ServiceAddress, node.ServicePort
+                    local svc_address, svc_port = nil, nil
+                    if node.Service then
+                        svc_address, svc_port = node.Service.Address, node.Service.Port
+                    end
```
Comment (Fabriceli) [inline apisix/discovery/consul/init.lua:273]: if `node.Service` is nil, it need to CONTINUE, do not insert in the `nodes` table, the node is empty

**Review by Fabriceli — CHANGES_REQUESTED** [review]

File: apisix/discovery/consul/init.lua (hunk 4 above)
Comment (dyrnq) [reply apisix/discovery/consul/init.lua:275]: update done

File: t/discovery/consul.t (hunk 3 above)
Comment (soulbird) [reply t/discovery/consul.t:582]: So， should delete this one？

Comment (dyrnq) [reply t/discovery/consul.t]: delete done

**Comment by dyrnq** [comment]
@Fabriceli Help me see what is left unfinished? There are some ci failed.

**Comment by Fabriceli** [comment]
> @Fabriceli Help me see what is left unfinished? There are some ci failed.

Please fix the test cases, there are some test failed. Referenced infos about how to run APISIX test:
1. [APISIX Running tests](https://apisix.apache.org/docs/apisix/building-apisix/#running-tests)
2. [How to Use Gitpod to Develop API Gateway?](https://apisix.apache.org/blog/2022/03/03/develop-apisix-with-gitpod/)

**Comment by dyrnq** [comment]
> > @Fabriceli Help me see what is left unfinished? There are some ci failed.
> 
> Please fix the test cases, there are some test failed. Referenced infos about how to run APISIX test:
> 
> 1. [APISIX Running tests](https://apisix.apache.org/docs/apisix/building-apisix/#running-tests)
> 2. [How to Use Gitpod to Develop API Gateway?](https://apisix.apache.org/blog/2022/03/03/develop-apisix-with-gitpod/)

Tks , let me learn and fix.

**Comment by dyrnq** [comment]
@Fabriceli there are some test use 8600 which is the first consul DNS port, it can not use to another consul instance(8500), i want to update it to 18500.

**Comment by dyrnq** [comment]
> @Fabriceli there are some test use 8600 which is the first consul DNS port, it can not use to another consul instance(8500), i want to update it to 18500.

I see there use docker-compose expose 8600 from 8500 , I will follow this first , there are so many changes.

**Comment by dyrnq** [comment]
@Fabriceli 

I encountered difficulties.

I notice the test ci use docker-compose start consul_1 8500 and consul_2 8600 and then they cant access `http://127.0.0.1:30511` which register to consul , the health check always check failed. That will not be expected.

**Comment by Fabriceli** [comment]
> @Fabriceli
> 
> I encountered difficulties.
> 
> I notice the test ci use docker-compose start consul_1 8500 and consul_2 8600 and then they can not access `http://127.0.0.1:30511` which register to consul , the health check always check failed. That will not be expected.
> 
> I need use host IP register service replace `127.0.0.1` , so consul (start by docker-compose) can check the register instance IP successfully

L26-43: https://github.com/apache/apisix/blob/master/t/discovery/consul_dump.t#L26

**Comment by dyrnq** [comment]
> > @Fabriceli
> > I encountered difficulties.
> > I notice the test ci use docker-compose start consul_1 8500 and consul_2 8600 and then they can not access `http://127.0.0.1:30511` which register to consul , the health check always check failed. That will not be expected.
> > I need use host IP register service replace `127.0.0.1` , so consul (start by docker-compose) can check the register instance IP successfully
> 
> L26-43: https://github.com/apache/apisix/blob/master/t/discovery/consul_dump.t#L26

Yes , I use here code ,I mean this instance reg Address can use 127.0.0.1:30511 but the checks can not use 127.0.0.1:30511
because the consul use docker ,it can`t check 127.0.0.1:30511 , 30511 on host, not in consul container ,check health are always failed, look the check http use `http://192.168.6.35:30511` can work fine, but `http://127.0.0.1:30511` not ok.

```bash
"PUT /v1/agent/service/register\n" . "{\"Checks\": [{\"http\": \"http://192.168.6.35:30511\",\"interval\": \"1s\"}],\"ID\":\"service_a1\",\"Name\":\"service_a\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":30511,\"Meta\":{\"service_a_version\":\"4.0\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",```

**Comment by dyrnq** [comment]
Or maybe I can use A fake address that can be visited instead?

```bash
 \"Checks\": [{\"http\": \"http://baidu.com\"
```

It can work

**Comment by dyrnq** [comment]
@Fabriceli 
I have unit test completed on my pc use flow test case, I have push the code. Please help checks.

```bash
=== TEST 16: prepare nodes with consul health check
--- config
location /v1/agent {
    proxy_pass http://127.0.0.1:8500;
}
--- request eval
[
    "PUT /v1/agent/service/deregister/service_a1",
    "PUT /v1/agent/service/deregister/service_a2",
    "PUT /v1/agent/service/deregister/service_b1",
    "PUT /v1/agent/service/deregister/service_b2",
    "PUT /v1/agent/service/register\n" . "{\"Checks\": [{\"http\": \"http://baidu.com\",\"interval\": \"1s\"}],\"ID\":\"service_a1\",\"Name\":\"service_a\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":30511,\"Meta\":{\"service_a_version\":\"4.0\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",
    "PUT /v1/agent/service/register\n" . "{\"Checks\": [{\"http\": \"http://127.0.0.1:8002\",\"interval\": \"1s\"}],\"ID\":\"service_b1\",\"Name\":\"service_b\",\"Tags\":[\"primary\",\"v1\"],\"Address\":\"127.0.0.1\",\"Port\":8002,\"Meta\":{\"service_b_version\":\"4.1\"},\"EnableTagOverride\":false,\"Weights\":{\"Passing\":10,\"Warning\":1}}",
]
--- response_body eval
--- error_code eval
[200, 200, 200, 200, 200, 200]


=== TEST 17: show dump services with consul health check
--- yaml_config
apisix:
  node_listen: 1984
  enable_control: true
discovery:
  consul:
    servers:
      - "http://127.0.0.1:8500"
    dump:
      path: "consul.dump"
      load_on_init: false
--- config
    location /t {
        content_by_lua_block {
            local json = require("toolkit.json")
            local t = require("lib.test_admin")
            ngx.sleep(2)

            local code, body, res = t.test('/v1/discovery/consul/show_dump_file',
                ngx.HTTP_GET)
            local entity = json.decode(res)
            ngx.say(json.encode(entity.services))
        }
    }
--- timeout: 3
--- request
GET /t
--- response_body
{"service_a":[{"host":"127.0.0.1","port":30511,"weight":1}]}
```
```
prove -v -I. -Itest-nginx/inc -Itest-nginx/lib -r t/discovery/consul_dump2.t


t/discovery/consul_dump2.t .. 
ok 1 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - status code ok
ok 2 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - pattern "[error]" does not match a line in error.log (req 0)
ok 3 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - status code ok
ok 4 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - pattern "[error]" does not match a line in error.log (req 0)
ok 5 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - status code ok
ok 6 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - pattern "[error]" does not match a line in error.log (req 0)
ok 7 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - status code ok
ok 8 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - pattern "[error]" does not match a line in error.log (req 0)
ok 9 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - status code ok
ok 10 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - pattern "[error]" does not match a line in error.log (req 0)
ok 11 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - status code ok
ok 12 - t/discovery/consul_dump2.t TEST 16: prepare nodes with consul health check - pattern "[error]" does not match a line in error.log (req 0)
ok 13 - t/discovery/consul_dump2.t TEST 17: show dump services with consul health check - status code ok
ok 14 - t/discovery/consul_dump2.t TEST 17: show dump services with consul health check - response_body - response is expected (repeated req 0, req 0)
ok 15 - t/discovery/consul_dump2.t TEST 17: show dump services with consul health check - pattern "[error]" does not match a line in error.log (req 0)
1..15
ok
All tests successful.
Files=1, Tests=15,  3 wallclock secs ( 0.02 usr  0.00 sys +  0.28 cusr  0.06 csys =  0.36 CPU)
Result: PASS

```

**Comment by Fabriceli** [comment]
cc @spacewander

**Comment by spacewander** [comment]
Could you fix the CI: https://github.com/apache/apisix/actions/runs/4280301682/jobs/7470288647?

**Comment by dyrnq** [comment]
> Could you fix the CI: https://github.com/apache/apisix/actions/runs/4280301682/jobs/7470288647?

I am in process of this.

**Comment by dyrnq** [comment]
@Fabriceli @spacewander 
I have fix , but there has a small discussion and uncertainty about a config of
```bash
discovery:
  consul:
    timeout:
      wait: 60
```
the wait query param in consul need s or m  obvious setting，so in code i need add .. "s"
```bash
wait = consul_conf.timeout.wait .."s", --blocked wait!=0; unblocked by wait=0
```

```bash
curl -fsSL -i http://127.0.0.1:8500/v1/health/state/any?wait=1s
HTTP/1.1 200 OK
Content-Type: application/json
Vary: Accept-Encoding
X-Consul-Effective-Consistency: leader
X-Consul-Index: 544
X-Consul-Knownleader: true
X-Consul-Lastcontact: 0
Date: Tue, 28 Feb 2023 04:11:07 GMT
Content-Length: 250

[{"Node":"8cd1fbccae10","CheckID":"serfHealth","Name":"Serf Health Status","Status":"passing","Notes":"","Output":"Agent alive and reachable","ServiceID":"","ServiceName":"","ServiceTags":[],"Type":"","Definition":{},"CreateIndex":5,"ModifyIndex":5}]


curl -fsSL -i http://127.0.0.1:8500/v1/health/state/any?wait=1
curl: (22) The requested URL returned error: 400 Bad Request
```

**Comment by Fabriceli** [comment]
please check it (**timeout** arg) in [lua-resty-consul](https://github.com/hamishforbes/lua-resty-consul), we use this lib to get infos from Consul

**Comment by Fabriceli** [comment]
Before push the code, you can run `make lint` and `make license-check` in local to check the code and doc whether fit the code style

**Comment by dyrnq** [comment]
> please check it (**wait** arg) in [lua-resty-consul](https://github.com/hamishforbes/lua-resty-consul), we use this lib to get infos from Consul

got it

**Comment by dyrnq** [comment]
@Fabriceli

Very strange question, the consul <http://127.0.0.1:8500/v1/health/state/any> update have delay when consul service node changed

when use **date && nerdctl stop nginx-1** health/state/any have 10s delay,
when use **date && nerdctl start nginx-1** health/state/any have 2s delay,


Create this script called consul_check_services_changes.sh
```bash
#!/bin/bash
if test ${1:-none} = "none"
then
  echo "USAGE: $0 Consul_URL"
  echo "       Example: localhost:8500/v1/health/service/MY_SUPER_SERVICE"
  exit 1
fi
url_to_check=$1

headers=$(mktemp)
content=$(mktemp)
index=0
while true;
do
  url="${url_to_check}?wait=5m&index=${index}&pretty=true&stale"
  curl -fs --dump-header "$headers" -o "${content}.new" "${url}" || { "echo Failed to query ${url}"; exit 1; }
  if test $index -ne 0
  then
    diff -u "$content" "$content.new" && echo " diff: No Differences found in service"
  fi
  index=$(grep "X-Consul-Index" "$headers" | sed 's/[^0-9]*\([0-9][0-9]*\)[^0-9]*/\1/g')
  mv "$content.new" "$content"
  printf "X-Consul-Index: $index at $(date) \b"
done
```
```bash
./consul_check_services_changes.sh http://127.0.0.1:8500/v1/health/state/any
```

If /health/state/any change has deplay it can not use to for comparison index change.I'm looking for a solution

**Comment by dyrnq** [comment]
@Fabriceli 

After continuous testing, I found problems.

If the consul register and deregister with no health check , then the  v1/health/state/any will no change of X-Consul-Index at all.

Be compatible with consul check and no check services ,we need to watch health/state/any Index and catalog/services index both, if any index change go into update process

**Comment by Fabriceli** [comment]
> 

we need create some tests for these cases

**Comment by dyrnq** [comment]
> > 
> 
> we need create some tests for these cases

I am not sure the resty_consul:new code how block query two API, if one block and wait, how another exec?

**Comment by Fabriceli** [comment]
> > > 
> > 
> > 
> > we need create some tests for these cases
> 
> I am not sure the resty_consul:new code how block two query API, if one block and wait, how another exec?
> 
> now use config keepalive false And remove the -- if current index different last index then update service compare,it will work fine
> 
> ```
>     fetch_interval: 1
>     keepalive: false
> ```
> 
> That is to say, long pooling connection is not used

that will be a BREAKING CHANGE

**Comment by dyrnq** [comment]
Yes, that lost keepalive usage.

We need find a compatible mechanism to watch catalog/services AND health/state/any X-Consul-Index both, 

I found all of consul APIs, there has no an API for sensing both service and health change.

**Comment by dyrnq** [comment]
Now I use commit code like below can pass test,  that is give up pass index parameter to consul , do not use blocking query, it will work fine.

Is there have a better way?

```bash
        -- consul_server.index = watch_result.headers['X-Consul-Index']
        -- -- only long connect type use index
        -- if consul_server.keepalive then
        --     consul_server.default_args.index = watch_result.headers['X-Consul-Index']
        -- end
```

**Comment by Fabriceli** [comment]
It can create TWO event to watch the services and update the `all_services`:

```mermaid
flowchart TD
		subgraph Event1: watch catalog
		A[Start] --> catalog{Catalog any change?}
    catalog -->|Yes| health[Query nodes]
    health --> D[Post Event1]
    catalog ---->|No| E[End]
    D --> E
    end
    
    subgraph Event2: watch health
    A2[Start] --> health2{health any change?}
    health2 -->|Yes| catalog2[Query services]
    catalog2 --> H[Query nodes]
    H --> D2[Post Event2]
    health2 ---->|No| E2[End]
    D2 --> E2
    end
    
    subgraph Events register
    A3[Start] --> register[Register Event]
    register --> update[Update all services]
    update --> e[End]
    end
```

**Comment by dyrnq** [comment]
I don't know much about this mechanism abhout events,I dare not move that piece of code. I'm afraid to fix more bugs.

 Can you join and help finish the work together.

**Comment by Fabriceli** [comment]
> I don't know much about this mechanism about events,I dare not move that piece of code. I'm afraid to fix more bugs.
> 
> Can you join and help finish the work together.

I create a new PR to fix this "bug", if you are agree

**Comment by dyrnq** [comment]
> > I don't know much about this mechanism about events,I dare not move that piece of code. I'm afraid to fix more bugs.
> > Can you join and help finish the work together.
> 
> I create a new PR to fix this "bug", if you are agree

Great ! agree with you

**Comment by monkeyDluffy6017** [comment]
Considered resolved due to https://github.com/apache/apisix/pull/9204

**Pull request closed without merging**